File: forms/dyslexia-assessment/full-stack-with-loco-tera-htmx-alpine/tests/engine/dyslexia_grader_test.rs
use dyslexia_assessment_tera_crate::engine::dyslexia_grader::grade;
use dyslexia_assessment_tera_crate::engine::dyslexia_rules::all_rules;
use dyslexia_assessment_tera_crate::engine::types::*;

/// Build an assessment with all nine domain scores at average (100).
fn average_case() -> AssessmentData {
    AssessmentData {
        reading_assessment: ReadingAssessment {
            reading_fluency_score: Some(100),
            reading_comprehension_score: Some(100),
            ..ReadingAssessment::default()
        },
        writing_spelling: WritingSpelling {
            spelling_accuracy_score: Some(100),
            written_expression_score: Some(100),
            ..WritingSpelling::default()
        },
        phonological_processing: PhonologicalProcessing {
            phonological_awareness_score: Some(100),
            phonological_memory_score: Some(100),
            rapid_naming_score: Some(100),
            ..PhonologicalProcessing::default()
        },
        working_memory_processing_speed: WorkingMemoryProcessingSpeed {
            working_memory_score: Some(100),
            processing_speed_score: Some(100),
            ..WorkingMemoryProcessingSpeed::default()
        },
        ..AssessmentData::default()
    }
}

#[test]
fn empty_assessment_is_none_severity() {
    let data = AssessmentData::default();
    let result = grade(&data);
    assert_eq!(result.overall_severity, "none");
    assert_eq!(result.answered_count, 0);
    assert!(result.lowest_score.is_none());
    assert_eq!(result.domain_scores.len(), 9);
}

#[test]
fn fully_average_case_is_none_severity() {
    let data = average_case();
    let result = grade(&data);
    assert_eq!(result.overall_severity, "none");
    assert_eq!(result.answered_count, 9);
    assert_eq!(result.lowest_score, Some(100));
}

#[test]
fn one_severe_score_drives_overall_severe() {
    let mut data = average_case();
    data.reading_assessment.reading_fluency_score = Some(50);
    let result = grade(&data);
    assert_eq!(result.overall_severity, "severe");
    assert_eq!(result.lowest_score, Some(50));
}

#[test]
fn moderate_score_drives_overall_moderate() {
    let mut data = average_case();
    data.phonological_processing.phonological_awareness_score = Some(60);
    let result = grade(&data);
    assert_eq!(result.overall_severity, "moderate");
    assert_eq!(result.lowest_score, Some(60));
}

#[test]
fn mild_score_drives_overall_mild() {
    let mut data = average_case();
    data.working_memory_processing_speed.processing_speed_score = Some(75);
    let result = grade(&data);
    assert_eq!(result.overall_severity, "mild");
    assert_eq!(result.lowest_score, Some(75));
}

#[test]
fn most_severe_band_wins() {
    let mut data = average_case();
    data.reading_assessment.reading_fluency_score = Some(75); // mild
    data.writing_spelling.spelling_accuracy_score = Some(60); // moderate
    data.phonological_processing.rapid_naming_score = Some(50); // severe
    let result = grade(&data);
    assert_eq!(result.overall_severity, "severe");
    assert_eq!(result.lowest_score, Some(50));
}

#[test]
fn rule_ids_are_unique_and_complete() {
    let rules = all_rules();
    assert_eq!(rules.len(), 9, "exactly 9 dyslexia rules expected");
    let mut ids: Vec<&str> = rules.iter().map(|r| r.id).collect();
    ids.sort();
    let len_before = ids.len();
    ids.dedup();
    assert_eq!(ids.len(), len_before, "rule IDs must be unique");
    for expected in [
        "DYS-001", "DYS-002", "DYS-003", "DYS-004", "DYS-005", "DYS-006", "DYS-007", "DYS-008",
        "DYS-009",
    ] {
        assert!(ids.contains(&expected), "missing rule ID {expected}");
    }
}

#[test]
fn domain_scores_include_all_rules() {
    let data = AssessmentData::default();
    let result = grade(&data);
    assert_eq!(result.domain_scores.len(), 9);
    for (i, ds) in result.domain_scores.iter().enumerate() {
        assert!(ds.id.starts_with("DYS-"), "domain {i} has bad id");
        assert_eq!(ds.severity, "none");
        assert!(ds.score.is_none());
    }
}

File: forms/dyslexia-assessment/front-end-with-svelte/src/lib/engine/dyslexia-grader.test.ts
import { describe, it, expect } from 'vitest';
import { gradeDyslexia } from './dyslexia-grader';
import { detectAdditionalFlags } from './flagged-issues';
import { dyslexiaRules } from './dyslexia-rules';
import { scoreSeverity, scoreBandLabel } from './utils';
import type { AssessmentData } from './types';

// A fresh, fully-blank assessment built inline. The store's
// createDefaultAssessment() is identical, but importing the store pulls in
// `$app/environment`, which is not resolvable in a plain Vitest run.
function base(): AssessmentData {
	return {
		demographics: {
			firstName: '',
			lastName: '',
			dateOfBirth: '',
			sex: '',
			ageYears: null,
			preferredLanguage: '',
			firstLanguage: '',
			handedness: '',
			referralSource: '',
			referralReason: ''
		},
		developmentalHistory: {
			pregnancyComplications: '',
			pregnancyDetails: '',
			birthComplications: '',
			birthDetails: '',
			earlyMilestones: '',
			speechDelay: '',
			languageDelay: '',
			hearingProblems: '',
			visionProblems: '',
			familyHistoryDyslexia: '',
			familyHistoryDetails: '',
			otherDevelopmentalNotes: ''
		},
		educationalBackground: {
			schoolType: '',
			currentYearGroup: '',
			schoolChanges: '',
			schoolChangeCount: null,
			attendanceIssues: '',
			attendanceDetails: '',
			eslLearner: '',
			academicStrengths: '',
			academicWeaknesses: '',
			previousAssessments: '',
			previousAssessmentDetails: ''
		},
		readingAssessment: {
			readingFluencyScore: null,
			readingComprehensionScore: null,
			difficultyDecoding: '',
			difficultyComprehension: '',
			avoidsReading: '',
			slowReadingSpeed: '',
			losesPlaceWhenReading: '',
			readingNotes: ''
		},
		writingSpelling: {
			spellingAccuracyScore: null,
			writtenExpressionScore: null,
			difficultySpelling: '',
			difficultyHandwriting: '',
			difficultyOrganisingIdeas: '',
			omitsLettersOrWords: '',
			reversesLettersOrNumbers: '',
			writingNotes: ''
		},
		phonologicalProcessing: {
			phonologicalAwarenessScore: null,
			phonologicalMemoryScore: null,
			rapidNamingScore: null,
			difficultyRhyming: '',
			difficultySegmentingSounds: '',
			difficultyBlendingSounds: '',
			difficultyLearningLetterSounds: '',
			phonologicalNotes: ''
		},
		workingMemoryProcessingSpeed: {
			workingMemoryScore: null,
			processingSpeedScore: null,
			difficultyFollowingInstructions: '',
			difficultyRememberingSequences: '',
			slowToCompleteTasks: '',
			difficultyTakingNotes: '',
			memoryNotes: ''
		},
		emotionalBehavioural: {
			lowSelfEsteem: '',
			anxietyAboutSchool: '',
			avoidanceBehaviour: '',
			frustrationWithLearning: '',
			peerRelationshipDifficulties: '',
			sleepDisturbance: '',
			mentalHealthConcerns: '',
			mentalHealthDetails: '',
			behaviouralNotes: ''
		},
		previousSupport: {
			previousIntervention: '',
			interventionTypes: '',
			currentEhcpOrIep: '',
			ehcpDetails: '',
			accessArrangements: '',
			accessArrangementsList: [],
			tutorialSupport: '',
			assistiveTechnologyUsed: '',
			assistiveTechnologyDetails: '',
			previousSupportNotes: ''
		},
		recommendationsSupportPlan: {
			recommendedSupports: [],
			structuredLiteracyRecommended: '',
			assistiveTechRecommended: '',
			extraTimeRecommended: '',
			specialistAssessmentRecommended: '',
			parentTrainingRecommended: '',
			keyGoals: '',
			reviewTimeframe: '',
			additionalRecommendations: ''
		}
	};
}

describe('scoreSeverity', () => {
	it('classifies standardised scores into bands', () => {
		expect(scoreSeverity(110)).toBe('none');
		expect(scoreSeverity(85)).toBe('none');
		expect(scoreSeverity(84)).toBe('mild');
		expect(scoreSeverity(70)).toBe('mild');
		expect(scoreSeverity(69)).toBe('moderate');
		expect(scoreSeverity(55)).toBe('moderate');
		expect(scoreSeverity(54)).toBe('severe');
		expect(scoreSeverity(null)).toBe('none');
	});

	it('labels score bands', () => {
		expect(scoreBandLabel(120)).toBe('Above average');
		expect(scoreBandLabel(100)).toBe('Average');
		expect(scoreBandLabel(75)).toBe('Below average');
		expect(scoreBandLabel(60)).toBe('Well below average');
		expect(scoreBandLabel(50)).toBe('Significantly below average');
		expect(scoreBandLabel(null)).toBe('');
	});
});

describe('Dyslexia Grading Engine', () => {
	it('returns no dyslexia when no scores are entered', () => {
		const result = gradeDyslexia(base());
		expect(result.overallSeverity).toBe('none');
		expect(result.lowestScore).toBeNull();
		expect(result.answeredCount).toBe(0);
		expect(result.domainScores).toHaveLength(dyslexiaRules.length);
	});

	it('returns no dyslexia for all-average scores', () => {
		const d = base();
		d.readingAssessment.readingFluencyScore = 100;
		d.writingSpelling.spellingAccuracyScore = 95;
		const result = gradeDyslexia(d);
		expect(result.overallSeverity).toBe('none');
		expect(result.answeredCount).toBe(2);
		expect(result.lowestScore).toBe(95);
	});

	it('returns mild for a below-average reading fluency score', () => {
		const d = base();
		d.readingAssessment.readingFluencyScore = 78;
		const result = gradeDyslexia(d);
		expect(result.overallSeverity).toBe('mild');
		expect(result.lowestScore).toBe(78);
	});

	it('returns moderate for a well-below-average score', () => {
		const d = base();
		d.phonologicalProcessing.phonologicalAwarenessScore = 62;
		const result = gradeDyslexia(d);
		expect(result.overallSeverity).toBe('moderate');
		expect(result.lowestScore).toBe(62);
	});

	it('drives overall severity off the lowest (most-impaired) score', () => {
		const d = base();
		d.readingAssessment.readingFluencyScore = 100;
		d.writingSpelling.spellingAccuracyScore = 50;
		const result = gradeDyslexia(d);
		expect(result.overallSeverity).toBe('severe');
		expect(result.lowestScore).toBe(50);
		expect(result.answeredCount).toBe(2);
	});

	it('keeps all rule IDs unique', () => {
		const ids = dyslexiaRules.map((r) => r.id);
		expect(new Set(ids).size).toBe(ids.length);
	});
});

describe('Dyslexia Flagged Issues Detection', () => {
	it('returns no flags for an empty assessment', () => {
		expect(detectAdditionalFlags(base())).toHaveLength(0);
	});

	it('flags a severe reading fluency score', () => {
		const d = base();
		d.readingAssessment.readingFluencyScore = 50;
		const flags = detectAdditionalFlags(d);
		expect(flags.some((f) => f.id === 'FLAG-READ-001')).toBe(true);
	});

	it('flags reported hearing problems', () => {
		const d = base();
		d.developmentalHistory.hearingProblems = 'yes';
		const flags = detectAdditionalFlags(d);
		expect(flags.some((f) => f.id === 'FLAG-DEV-001')).toBe(true);
	});

	it('flags family history of dyslexia', () => {
		const d = base();
		d.developmentalHistory.familyHistoryDyslexia = 'yes';
		const flags = detectAdditionalFlags(d);
		expect(flags.some((f) => f.id === 'FLAG-DEV-005')).toBe(true);
	});

	it('flags mental health concerns as high priority', () => {
		const d = base();
		d.emotionalBehavioural.mentalHealthConcerns = 'yes';
		const flags = detectAdditionalFlags(d);
		const flag = flags.find((f) => f.id === 'FLAG-EMOT-004');
		expect(flag?.priority).toBe('high');
	});

	it('sorts flags by priority (high first)', () => {
		const d = base();
		d.readingAssessment.readingFluencyScore = 50; // high
		d.writingSpelling.reversesLettersOrNumbers = 'yes'; // low
		d.developmentalHistory.speechDelay = 'yes'; // medium
		const flags = detectAdditionalFlags(d);
		const order = { urgent: 0, high: 1, medium: 2, low: 3 };
		const priorities = flags.map((f) => order[f.priority]);
		const sorted = [...priorities].sort((a, b) => a - b);
		expect(priorities).toEqual(sorted);
	});
});
